src/kafka/kafka_clients.clj

```diff
@@ -0,0 +1,170 @@
+
+(comment
+
+  ;; build producer config
+  (do (def producer-config (build-producer-config-atom))
+      producer-config)
+  #_=> #_#object[clojure.lang.Atom
+                 0x2d4203fd
+                 {:status :ready,
+                  :val    {"value.serializer"  org.apache.kafka.common.serialization.StringSerializer,
+                           "key.serializer"    org.apache.kafka.common.serialization.StringSerializer,
+                           "client.id"         "m-w1kx3xw0nr",
+                           "bootstrap.servers" "localhost:9092",
+                           "acks"              "all"}}]
+
+  ;; build consumer config
+  (do (def consumer-config (build-consumer-config-atom))
+      consumer-config)
+  #_=> #_#object[clojure.lang.Atom
+                 0xf91e958
+                 {:status :ready,
+                  :val    {"group.id"           "My-Group",
+                           "key.deserializer"   org.apache.kafka.common.serialization.StringDeserializer,
+                           "value.deserializer" org.apache.kafka.common.serialization.StringDeserializer,
+                           "client.id"          "m-w1kx3xw0nr",
+                           "bootstrap.servers"  "localhost:9092"}}]
+
+  ;; build producer
+  (do (def producer (build-producer-atom producer-config))
+      producer)
+  #_=> #_#object[clojure.lang.Atom
+                 0x26014103
+                 {:status :ready,
+                  :val    #object[org.apache.kafka.clients.producer.KafkaProducer
+                                  0x1325259b
+                                  "org.apache.kafka.clients.producer.KafkaProducer@1325259b"]}]
+
+  ;; build consumer and subscribe topics
+  (do (def consumer (build-consumer-atom consumer-config))
+      consumer)
+  #_=> #_#object[clojure.lang.Atom
+                 0x1c8b04b8
+                 {:status :ready,
+                  :val    #object[org.apache.kafka.clients.consumer.KafkaConsumer
+                                  0x6213626d
+                                  "org.apache.kafka.clients.consumer.KafkaConsumer@6213626d"]}]
+
+  ;; generate record
+  (do (def record (new ProducerRecord "demo" "Hello world!"))
+      record)
+  #_=> #_#object[org.apache.kafka.clients.producer.ProducerRecord
+                 0x3fb54ba6
+                 "ProducerRecord(topic=demo, partition=null, headers=RecordHeaders(headers = [], isReadOnly = false), key=null, value=Hello world!, timestamp=null)"]
+
+  ;; producer send record
+  (send-msg! producer record)
+  #_=> #_#object[org.apache.kafka.clients.producer.internals.FutureRecordMetadata
+                 0xde5318c
+                 "org.apache.kafka.clients.producer.internals.FutureRecordMetadata@de5318c"]
+
+  ;; consumer receive message
+  (read-msg! consumer)
+  #_=> ["Processed Value: Hello world!"]
+
+
+
+  ;; if the producer/consumer died, we should rebuild them
+  ;; if we want to change the config, we should rebuild the consumer and producer
```

Comment:
Here are some scenarios where I think atom can be used please leave your review if there is anything wrong.